### PR TITLE
Push parser

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Introduce a new SAX-style JSON push parser and integrate it into the picojson library, extend error handling, and add extensive conformance and stress tests.

New Features:
- Add PushParser with PushParserHandler trait for event-driven JSON parsing
- Expose PushParser, PushParseError, and PushParserHandler in the public API

Bug Fixes:
- Use f64::EPSILON for float comparison in StreamParser tests

Enhancements:
- Expand ParseError with Utf8 and InputBufferFull variants and unify ujson::Error conversion
- Optimize escape handling in PushParser with copy-on-escape design

Tests:
- Integrate PushParser and StreamParser into existing JSON checker pass/fail tests
- Add new test modules for push parser conformance, stress (buffer sizes and chunk patterns), unicode and escape edge cases, copy-on-escape behavior, and invalid slice bounds reproduction

## Summary by CodeRabbit

* **New Features**
  * Introduced a new incremental (push) JSON parser with event-driven handling for streaming or chunked input.
  * Added a public push parser handler trait and expanded error types for parsing, handler, buffer, and UTF-8 errors.
  * Extended the public API to include push parser components and improved error conversions.

* **Bug Fixes**
  * Enhanced error detection and reporting for invalid or incomplete Unicode escape sequences and escape handling in keys and strings.
  * Tightened floating-point comparison precision in existing parser tests.

* **Tests**
  * Added extensive tests covering push parser functionality, escape sequences, Unicode handling, buffer boundary issues, and stress testing.
  * Improved conformance tests to cover all parser variants including push and stream parsers.
  * Introduced diagnostic tests reproducing buffer boundary issues and verifying string borrowing versus unescaping behavior.
  * Added tests for copy-on-escape optimizations and detailed event tracing for all JSON event types.

### CODE REVIEW

File: picojson/src/push_parser.rs [hunk 1]
```diff
@@ -0,0 +1,664 @@
...
+
+    /// Processes a chunk of input data.
+    pub fn write<'input, E>(&mut self, data: &'input [u8]) -> Result<(), PushParseError<E>>
+    where
+        H: for<'a, 'b> PushParserHandler<'a, 'b, E>,
+    {
+        // Event storage for tokenizer output. Fixed size of 2 events is by design:
+        // the tokenizer never returns more than 2 events per input byte. This is a 
+        // fundamental limit that all parsers in this codebase use.
+        let mut event_storage: [Option<(ujson::Event, usize)>; 2] = [None, None];
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs:79]: okay check please

File: picojson/src/push_parser.rs [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+                    }
+                    EscapeState::InEscapeSequence => {
+                        // Other escape sequences are handled by events, not by byte processing
+                    }
+                }
+            }
+        }
+
+        if let ParserState::ParsingString | ParserState::ParsingKey | ParserState::ParsingNumber =
+            self.state
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs:165]: doublecheck please

File: picojson/src/push_parser.rs [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+                            }
+                            Err(e) => return Err(e.into()),
+                        }
+                    }
+                }
+
+                // Process the collected unicode escape (now this will just reset the collector)
+                self.process_collected_unicode_escape()?;
+                self.escape_state = EscapeState::None;
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs]: addressed

File: picojson/src/push_parser.rs [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+                    | ujson::Event::End(ujson::EventToken::EscapeNewline)
+                    | ujson::Event::End(ujson::EventToken::EscapeCarriageReturn)
+                    | ujson::Event::End(ujson::EventToken::EscapeTab)
+                    | ujson::Event::End(ujson::EventToken::UnicodeEscape) => {
+                        should_append = false; // Don't append the escape trigger byte to buffer
+                        self.append_escape_content(event, pos, data)?
+                    }
+                    _ => self.append_escape_content(event, pos, data)?,
+                }
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs]: addressed

File: picojson/tests/json_checker_tests.rs [hunk 5]
```diff
@@ -85,6 +155,173 @@ mod json_checker_tests {
                 result.err()
             );
         }
+
+        // Debug test for tracing SliceParser unicode escape processing
+        #[test]
+        fn test_slice_parser_unicode_trace() {
+            // Test the problematic sequence from pass1.json line 45
+            let unicode_sequence = r#""\/\\\"\\uCAFE\\uBABE\\uAB98\\uFCDE\\ubcda\\uef4A\\b\\f\\n\\r\\t`1~!@#$%^&*()_+-=[]{}|;:',./<>?""#;
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/json_checker_tests.rs:150]: i'm almost sure this is a misunderstanding. The test works and passes

File: picojson/src/push_parser.rs [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+                self.stream_buffer.append_unescaped_byte(b'\r')?;
+                self.escape_state = EscapeState::None;
+            }
+            ujson::Event::End(ujson::EventToken::EscapeTab) => {
+                if !self.using_unescaped_buffer {
+                    self.switch_to_unescaped_mode(data, pos)?;
+                }
+                self.stream_buffer.append_unescaped_byte(b'\t')?;
+                self.escape_state = EscapeState::None;
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs]: handled

File: picojson/tests/json_checker_tests.rs [hunk 7]
```diff
@@ -85,6 +155,173 @@
...
+                event_count
+            );
+        }
+
+        // Debug test for tracing StreamParser unicode escape processing
+        #[test]
+        fn test_stream_parser_unicode_trace() {
+            // Test the problematic sequence from pass1.json line 45
+            let unicode_sequence = r#""\/\\\"\\uCAFE\\uBABE\\uAB98\\uFCDE\\ubcda\\uef4A\\b\\f\\n\\r\\t`1~!@#$%^&*()_+-=[]{}|;:',./<>?""#;
+
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/tests/json_checker_tests.rs:297]: Same misunderstanding as above

File: picojson/src/push_parser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_parser.rs:79]: Okay but what about this comment here ?

File: picojson/src/stream_parser.rs [hunk 8]
```diff
@@ -756,7 +756,7 @@ mod tests {
                     #[cfg(feature = "float")]
                     NumberResult::Float(f) => {
                         // This is expected in float-enabled build
-                        assert!((f - 3.14).abs() < 0.01);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/stream_parser.rs:759]: You seem confused - the code is already updated
